### PR TITLE
[KYUUBI #3064] Fix scala NPE issue when adding non-local jar URI to class loader

### LINKED ISSUE #3064: [Bug] Execute scala - NPE issue
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

```
11:16:39.018 INFO org.apache.kyuubi.operation.ExecuteStatement: Processing b_dapgap's query[d74ec33d-eedb-4f21-95a6-b7a4b067f004]: ERROR_STATE -> CLOSED_STATE, statement: spark.sparkContext.getConf.getOption
(
"spark.org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter.param.PROXY_URI_BASES")
.orElse(sc.uiWebUrl).getOrElse("").split(",").head
11:16:39.119 ERROR org.apache.kyuubi.server.KyuubiTBinaryFrontendService: Error executing statement:
org.apache.kyuubi.KyuubiSQLException: Error operating ExecuteScala: java.lang.NullPointerException
        at scala.tools.nsc.classpath.FileUtils$AbstractFileOps$.isJarOrZip$extension(FileUtils.scala:32)
        at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:91)
        at scala.tools.nsc.Global.$anonfun$extendCompilerClassPath$1(Global.scala:825)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38)
        at scala.collection.TraversableLike.map(TraversableLike.scala:238)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
        at scala.collection.AbstractTraversable.map(Traversable.scala:108)
        at scala.tools.nsc.Global.extendCompilerClassPath(Global.scala:825)
        at scala.tools.nsc.interpreter.IMain.addUrlsToClassPath(IMain.scala:259)
        at org.apache.kyuubi.engine.spark.operation.ExecuteScala.$anonfun$runInternal$1(ExecuteScala.scala:68)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.kyuubi.engine.spark.operation.SparkOperation.withLocalProperties(SparkOperation.scala:87)
        at org.apache.kyuubi.engine.spark.operation.ExecuteScala.runInternal(ExecuteScala.scala:60)
        at org.apache.kyuubi.operation.AbstractOperation.run(AbstractOperation.scala:164)
        at org.apache.kyuubi.session.AbstractSession.runOperation(AbstractSession.scala:97)
        at org.apache.kyuubi.engine.spark.session.SparkSessionImpl.runOperation(SparkSessionImpl.scala:76)
        at org.apache.kyuubi.session.AbstractSession.$anonfun$executeStatement$1(AbstractSession.scala:126)
        at org.apache.kyuubi.session.AbstractSession.withAcquireRelease(AbstractSession.scala:78)
        at org.apache.kyuubi.session.AbstractSession.executeStatement(AbstractSession.scala:123)
        at org.apache.kyuubi.service.AbstractBackendService.executeStatement(AbstractBackendService.scala:66)
        at org.apache.kyuubi.service.TFrontendService.ExecuteStatement(TFrontendService.scala:261)
        at org.apache.kyuubi.shade.org.apache.hive.service.rpc.thrift.TCLIService$Processor$ExecuteStatement.getResult(TCLIService.java:1557)
        at org.apache.kyuubi.shade.org.apache.hive.service.rpc.thrift.TCLIService$Processor$ExecuteStatement.getResult(TCLIService.java:1542)
        at org.apache.kyuubi.shade.org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
        at org.apache.kyuubi.shade.org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
        at org.apache.kyuubi.service.authentication.TSetIpAddressProcessor.process(TSetIpAddressProcessor.scala:36)
        at org.apache.kyuubi.shade.org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:286)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

        at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:69) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.engine.spark.operation.SparkOperation$$anonfun$onError$1.applyOrElse(SparkOperation.scala:112) ~[?:?]
        at org.apache.kyuubi.engine.spark.operation.SparkOperation$$anonfun$onError$1.applyOrElse(SparkOperation.scala:96) ~[?:?]
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38) ~[scala-library-2.12.15.jar:?]
        at org.apache.kyuubi.engine.spark.operation.ExecuteScala.$anonfun$runInternal$1(ExecuteScala.scala:88) ~[?:?]
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23) ~[scala-library-2.12.15.jar:?]
        at org.apache.kyuubi.engine.spark.operation.SparkOperation.withLocalProperties(SparkOperation.scala:87) ~[?:?]
        at org.apache.kyuubi.engine.spark.operation.ExecuteScala.runInternal(ExecuteScala.scala:60) ~[?:?]
        at org.apache.kyuubi.operation.AbstractOperation.run(AbstractOperation.scala:164) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.session.AbstractSession.runOperation(AbstractSession.scala:97) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.engine.spark.session.SparkSessionImpl.runOperation(SparkSessionImpl.scala:76) ~[?:?]
at org.apache.kyuubi.session.AbstractSession.$anonfun$executeStatement$1(AbstractSession.scala:126) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.session.AbstractSession.withAcquireRelease(AbstractSession.scala:78) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.session.AbstractSession.executeStatement(AbstractSession.scala:123) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.service.AbstractBackendService.executeStatement(AbstractBackendService.scala:66) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.service.TFrontendService.ExecuteStatement(TFrontendService.scala:261) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.shade.org.apache.hive.service.rpc.thrift.TCLIService$Processor$ExecuteStatement.getResult(TCLIService.java:1557) ~[?:?]
        at org.apache.kyuubi.shade.org.apache.hive.service.rpc.thrift.TCLIService$Processor$ExecuteStatement.getResult(TCLIService.java:1542) ~[?:?]
        at org.apache.kyuubi.shade.org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39) ~[?:?]
        at org.apache.kyuubi.shade.org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39) ~[?:?]
        at org.apache.kyuubi.service.authentication.TSetIpAddressProcessor.process(TSetIpAddressProcessor.scala:36) ~[kyuubi-common_2.12-1.6.0.1.5.0.jar:1.6.0.1.5.0]
        at org.apache.kyuubi.shade.org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:286) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_202]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_202]
        at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_202]
Caused by: java.lang.RuntimeException: java.lang.NullPointerException:null
        at scala.tools.nsc.classpath.FileUtils$AbstractFileOps$.isJarOrZip$extension(FileUtils.scala:32) ~[?:?]
        at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:91) ~[?:?]
        at scala.tools.nsc.Global.$anonfun$extendCompilerClassPath$1(Global.scala:825) ~[?:?]
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238) ~[scala-library-2.12.15.jar:?]
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36) ~[scala-library-2.12.15.jar:?]
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33) ~[scala-library-2.12.15.jar:?]
        at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38) ~[scala-library-2.12.15.jar:?]
        at scala.collection.TraversableLike.map(TraversableLike.scala:238) ~[scala-library-2.12.15.jar:?]
        at scala.collection.TraversableLike.map$(TraversableLike.scala:231) ~[scala-library-2.12.15.jar:?]
        at scala.collection.AbstractTraversable.map(Traversable.scala:108) ~[scala-library-2.12.15.jar:?]
        at scala.tools.nsc.Global.extendCompilerClassPath(Global.scala:825) ~[?:?]
        at scala.tools.nsc.interpreter.IMain.addUrlsToClassPath(IMain.scala:259) ~[?:?]
        at org.apache.kyuubi.engine.spark.operation.ExecuteScala.$anonfun$runInternal$1(ExecuteScala.scala:68) ~[?:?]
```

### Affects Version(s)

1.6.0

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Close #3064 

Refer the comments by @cxzl25 

Spark uses the code
 `URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory(hadoopConf))` supports the class path from HDFS.
But because the scala compiler only supports adding file schema urls to the class path, non-file schema urls will cause NPE.

```java
Error: Error operating ExecuteScala: java.lang.NullPointerException
	at scala.tools.nsc.classpath.FileUtils$AbstractFileOps$.isJarOrZip$extension(FileUtils.scala:32)
	at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:90)
	at scala.tools.nsc.Global.$anonfun$extendCompilerClassPath$1(Global.scala:832)
```

scala.tools.nsc.Global#extendCompilerClassPath
```scala
AbstractFile.getURL(u) 
```

scala.reflect.io.AbstractFile#getURL
```scala
  def getURL(url: URL): AbstractFile =
    if (url.getProtocol == "file") {
      val f = new java.io.File(url.toURI)
      if (f.isDirectory) getDirectory(f)
      else getFile(f)
    } else null
```

spark-shell supports --jars hdfs jar. At this time, submit will download the remote jar to the local and pass it to spark-shell through the `spark.repl.local.jars` configuration.

In this pr, I localize the remote jar url at first, and then add it into repl class path.
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala [hunk 1]
```diff
@@ -64,8 +68,19 @@ class ExecuteScala(
...
         warn(s"Clearing legacy output from last interpreting:\n $legacyOutput")
       }
-      val jars = spark.sharedState.jarClassLoader.getURLs
-      repl.addUrlsToClassPath(jars: _*)
+      spark.sharedState.jarClassLoader.getURLs.foreach { jar =>
+        try {
+          if ("file".equals(jar.toURI.getScheme)) {
+            repl.addUrlsToClassPath(jar)
+          } else {
+            spark.sparkContext.addFile(jar.toString)
```
Comment (turboFei) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala:76]: it is tricky, but I can not find another solution for this.

cc @yaooqinn  @iodone @942011334

**Comment by turboFei** [comment]
It seems that the RC is that:

for scala mode, we can not add jar dynamically.

**Comment by turboFei** [comment]
> It seems that the RC is that:
> 
> for scala mode, we can not add jar dynamically.
> 
> we can not add jar with `spark.sql("add jar $jar")` now.

image

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala (hunk 1 above)
Comment (iodone) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala:76]: `SparkSession.sessionState.resourceLoader.addJar` : 
image
There is a difference between the implementation of SparkContext.addJar` and `SparkSession.sessionState.resourceLoader.addJar`. `sparkcontext.add jar` does not have a jar path to add to the `classpath` of the current thread context

Comment (iodone) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala:76]: Why do we need to call `sparkContext.addjar` repeatedly here?

**Comment by yaooqinn** [comment]
@turboFei we have a test for add jar cmd?

> scala code with loading external package at runtime

**Comment by cxzl25** [comment]
While Spark uses the code
 `URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory(hadoopConf))` supports the class path from HDFS.
But because the scala compiler only supports adding file schema urls to the class path, non-file schema urls will cause NPE.



```java
Error: Error operating ExecuteScala: java.lang.NullPointerException
	at scala.tools.nsc.classpath.FileUtils$AbstractFileOps$.isJarOrZip$extension(FileUtils.scala:32)
	at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:90)
	at scala.tools.nsc.Global.$anonfun$extendCompilerClassPath$1(Global.scala:832)
```

scala.tools.nsc.Global#extendCompilerClassPath
```scala
AbstractFile.getURL(u) 
```

scala.reflect.io.AbstractFile#getURL
```scala
  def getURL(url: URL): AbstractFile =
    if (url.getProtocol == "file") {
      val f = new java.io.File(url.toURI)
      if (f.isDirectory) getDirectory(f)
      else getFile(f)
    } else null
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala (hunk 1 above)
Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala:76]: the jar url is in remote FS and scala repl can not identify that.

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteScala.scala:76]: the jar url is in remote FS and scala repl can not identify that.

**Comment by turboFei** [comment]
> @turboFei we have a test for add jar cmd?
> 
> >

Not yet. I tested it locally.

**Comment by turboFei** [comment]
how do you think about? @yaooqinn @pan3793 @cxzl25 @iodone

**Comment by cxzl25** [comment]
> how do you think about? 

We should avoid calling `repl.addUrlsToClassPath` every execution, which will lead to more and more urls in the `classloaderURLs` of `ScalaClassLoader`, and `URLClassPath#addURL` will also check whether the url exists or not every time.

Other changes LGTM.


scala.reflect.internal.util.ScalaClassLoader.URLClassLoader#addURL
```scala
    override def addURL(url: URL) = {
      classloaderURLs :+= url
      super.addURL(url)
    }
```
jdk.internal.loader.URLClassPath#addURL
```java
    if (! path.contains(url)) {
```

**Comment by turboFei** [comment]
thanks @cxzl25 , resolved.

**Review by cxzl25 — APPROVED** [review]

**Comment by turboFei** [comment]
thanks a lot for the review. @cxzl25 @iodone @yaooqinn 

merging to master

**Comment by pan3793** [comment]
This looks like a bugfix, should we backport to branch-1.6?

**Comment by pan3793** [comment]
and please fill the milestone when closing PR.

**Comment by turboFei** [comment]
thanks, merged to branch-1.6 as well.

**Pull request closed without merging**